Flake8 errors

`flake8==3.8.1` catches a few new flake8 errors that are making CI tests fail:
```
.\src\hdmf\container.py:342:26: E741 ambiguous variable name 'l'
.\src\hdmf\backends\hdf5\h5tools.py:790:5: C901 'HDF5IO.write_dataset' is too complex (31)
.\src\hdmf\backends\hdf5\h5tools.py:1182:29: E741 ambiguous variable name 'l'
.\src\hdmf\backends\hdf5\h5_utils.py:392:69: E225 missing whitespace around operator
.\src\hdmf\build\builders.py:206:13: E741 ambiguous variable name 'l'
.\src\hdmf\common\table.py:201:5: C901 'DynamicTable.__init__' is too complex (32)
.\src\hdmf\common\table.py:558:23: F507 '...' % ... has 1 placeholder(s) but 2 substitution(s)
.\src\hdmf\validate\validator.py:402:5: C901 'GroupValidator.validate' is too complex (21)
.\tests\coloredtestrunner.py:94:26: E741 ambiguous variable name 'l'
.\tests\unit\test_io_hdf5.py:24:17: E722 do not use bare 'except'
.\tests\unit\spec_tests\test_spec_write.py:54:1: E122 continuation line missing indentation or outdented
.\tests\unit\spec_tests\test_spec_write.py:71:1: E122 continuation line missing indentation or outdented
.\tests\unit\spec_tests\test_spec_write.py:200:1: E122 continuation line missing indentation or outdented
```

--- a/src/hdmf/backends/hdf5/h5_utils.py
+++ b/src/hdmf/backends/hdf5/h5_utils.py
@@ -389,7 +389,7 @@ class H5DataIO(DataIO):
             )
     def __init__(self, **kwargs):
         # Get the list of I/O options that user has passed in
-        ioarg_names = [name for name in kwargs.keys() if name not in['data', 'link_data', 'allow_plugin_filters']]
+        ioarg_names = [name for name in kwargs.keys() if name not in ['data', 'link_data', 'allow_plugin_filters']]
         # Remove the ioargs from kwargs
         ioarg_values = [popargs(argname, kwargs) for argname in ioarg_names]
         # Consume link_data parameter

--- a/src/hdmf/backends/hdf5/h5_utils.py
+++ b/src/hdmf/backends/hdf5/h5_utils.py
@@ -273,7 +273,7 @@ class H5SpecWriter(SpecWriter):
     def __write(self, d, name):
         data = self.stringify(d)
         # create spec group if it does not exist. otherwise, do not overwrite existing spec
-        dset = self.__group.require_dataset(name, shape=tuple(), data=data, dtype=self.__str_type)
+        dset = self.__group.create_dataset(name, shape=tuple(), data=data, dtype=self.__str_type)
         return dset
 
     def write_spec(self, spec, path):

--- a/src/hdmf/backends/hdf5/h5tools.py
+++ b/src/hdmf/backends/hdf5/h5tools.py
@@ -1161,9 +1161,9 @@ class HDF5IO(HDMFIO):
         # dependency
         self.__ref_queue.append(func)
 
-    def __rec_get_ref(self, l):
+    def __rec_get_ref(self, ref_list):
         ret = list()
-        for elem in l:
+        for elem in ref_list:
             if isinstance(elem, (list, tuple)):
                 ret.append(self.__rec_get_ref(elem))
             elif isinstance(elem, (Builder, Container)):

--- a/src/hdmf/backends/hdf5/h5tools.py
+++ b/src/hdmf/backends/hdf5/h5tools.py
@@ -294,7 +294,9 @@ class HDF5IO(HDMFIO):
                 ns_builder = self.__convert_namespace(ns_catalog, ns_name)
                 namespace = ns_catalog.get_namespace(ns_name)
                 group_name = '%s/%s' % (ns_name, namespace.version)
-                ns_group = spec_group.require_group(group_name)
+                if group_name in spec_group:
+                    continue
+                ns_group = spec_group.create_group(group_name)
                 writer = H5SpecWriter(ns_group)
                 ns_builder.export('namespace', writer=writer)
 

--- a/src/hdmf/build/builders.py
+++ b/src/hdmf/build/builders.py
@@ -197,15 +197,15 @@ class GroupBuilder(BaseBuilder):
         source when this source is set
         '''
         super(GroupBuilder, self.__class__).source.fset(self, s)
-        for g in self.groups.values():
-            if g.source is None:
-                g.source = s
-        for d in self.datasets.values():
-            if d.source is None:
-                d.source = s
-        for l in self.links.values():
-            if l.source is None:
-                l.source = s
+        for group in self.groups.values():
+            if group.source is None:
+                group.source = s
+        for dset in self.datasets.values():
+            if dset.source is None:
+                dset.source = s
+        for link in self.links.values():
+            if link.source is None:
+                link.source = s
 
     @property
     def groups(self):

--- a/src/hdmf/common/table.py
+++ b/src/hdmf/common/table.py
@@ -555,7 +555,7 @@ class DynamicTable(Container):
         region = getargs('region', kwargs)
         if isinstance(region, slice):
             if (region.start is not None and region.start < 0) or (region.stop is not None and region.stop > len(self)):
-                msg = 'region slice %s is out of range for this DynamicTable of length ' % (str(region), len(self))
+                msg = 'region slice %s is out of range for this DynamicTable of length %d' % (str(region), len(self))
                 raise IndexError(msg)
             region = list(range(*region.indices(len(self))))
         else:

--- a/src/hdmf/container.py
+++ b/src/hdmf/container.py
@@ -339,20 +339,20 @@ class Container(AbstractContainer):
             return str(v)
 
     @staticmethod
-    def __smart_str_list(l, num_indent, left_br):
+    def __smart_str_list(str_list, num_indent, left_br):
         if left_br == '(':
             right_br = ')'
         if left_br == '{':
             right_br = '}'
-        if len(l) == 0:
+        if len(str_list) == 0:
             return left_br + ' ' + right_br
         indent = num_indent * 2 * ' '
         indent_in = (num_indent + 1) * 2 * ' '
         out = left_br
-        for v in l[:-1]:
+        for v in str_list[:-1]:
             out += '\n' + indent_in + Container.__smart_str(v, num_indent + 1) + ','
-        if l:
-            out += '\n' + indent_in + Container.__smart_str(l[-1], num_indent + 1)
+        if str_list:
+            out += '\n' + indent_in + Container.__smart_str(str_list[-1], num_indent + 1)
         out += '\n' + indent + right_br
         return out
 

--- a/tests/coloredtestrunner.py
+++ b/tests/coloredtestrunner.py
@@ -91,7 +91,7 @@ class Table(object):
         return len(re.sub(r"\033\[[0-9];[0-9];[0-9]{1,2}m", "", x))
 
     def addRow(self, row):
-        rows = [[''] for l in range(len(row))]
+        rows = [[''] for i in range(len(row))]
         maxrows = 0
         for i, x in enumerate(row):
             for j, y in enumerate(x.split("\n")):

--- a/tests/unit/spec_tests/test_spec_write.py
+++ b/tests/unit/spec_tests/test_spec_write.py
@@ -61,7 +61,7 @@ class TestSpec(TestCase):
   - name: testdata
     dtype: float
     doc: test
-"""  # noqa: E128
+"""  # noqa: E122
             nsstr = file.read()
             self.assertEqual(nsstr, match_str)
 
@@ -80,7 +80,7 @@ class TestSpec(TestCase):
     source: mylab.extensions.yaml
     title: Extensions for my lab
   version: 0.0.1
-""" % self.date.isoformat()  # noqa: E128
+""" % self.date.isoformat()  # noqa: E122
             nsstr = file.read()
             self.assertEqual(nsstr, match_str)
 
@@ -207,7 +207,7 @@ class TestExportSpec(TestSpec):
   schema:
   - source: mylab.extensions.yaml
   version: 0.0.1
-""" % self.date.isoformat()  # noqa: E128
+""" % self.date.isoformat()  # noqa: E122
             nsstr = file.read()
             self.assertEqual(nsstr, match_str)
 

--- a/tests/unit/test_io_hdf5.py
+++ b/tests/unit/test_io_hdf5.py
@@ -21,7 +21,7 @@ class HDF5Encoder(json.JSONEncoder):
                 try:
                     ret = t(obj)
                     break
-                except:  # noqa: F722
+                except:  # noqa: E722
                     pass
             if ret is None:
                 return obj

